### ISSUE TITLE
Just a question about the fancy graph!

### ISSUE THREAD

**Issue by bartekpacia** [buildkite/agent#2932]
Hey folks @ Buildkite,

I noticed [this great graph](https://github.com/buildkite/agent/blob/952eaef0c6d84b7c87097966f3309a3255fc0cc9/docs/agent-start.md) in the docs, and can't resist asking how did you generate create the `.dot` file for it? Some tool that traces function calls, or just meticulous work?

Thanks!

**Comment by DrJosh9000** [comment]
Hi @bartekpacia ! Thanks for the question. Meticulous work, basically.

Not long after I started with Buildkite, I had concerns about the number and lifetime of all the various goroutines created between starting the agent and running a job. With that in mind, I ended up manually working through each function call and goroutine, and it felt natural to start keeping notes in the form of a graph. (And then publish it in case anyone was interested.)

I would be interested in an automated tool for doing it though! I imagine extracting call graphs from runtime traces or `pprof` profiles would be tractable.
